src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,640 @@
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        _cleanup_free_ char *res = NULL;
+        char **out = data;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        /* XDG does not allow duplicate definitions. */
+        if (*out) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Key %s was defined multiple times, ignoring.", lvalue);
+                return 0;
+        }
+
+        res = strdup(skip_leading_chars(rvalue, NULL));
+        if (!res)
+                return log_oom();
+
+        r = xdg_unescape_string(unit, filename, line, res);
+        if (r < 0)
+                return r;
+
+        *out = TAKE_PTR(res);
+        return 0;
+}
+
+static int xdg_config_parse_strv(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        char ***sv = data;
+        char *start;
+        char *end;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        /* XDG does not allow duplicate definitions. */
+        if (*sv) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Key %s was defined multiple times, ignoring.", lvalue);
+                return 0;
+        }
+
+        *sv = strv_new(NULL);
+        if (!*sv)
+                return log_oom();
+
+        /* We cannot use strv_split because it does not handle escaping correctly. */
+        start = skip_leading_chars(rvalue, NULL);
+
+        for (end = start; *end; end++) {
+                if (*end == '\\') {
+                        /* Move forward, and ensure it is a valid escape. */
+                        end++;
+                        if (strchr("sntr\\;", *end) == NULL) {
+                                log_syntax(unit, LOG_ERR, filename, line, 0, "Undefined escape sequence \\%c.", *end);
+                                return 0;
+                        }
+                        continue;
+                }
+
+                if (*end == ';') {
+                        _cleanup_free_ char *copy = NULL;
+
+                        copy = strndup(start, end - start);
+                        if (copy == NULL)
+                                return log_oom();
+                        r = xdg_unescape_string(unit, filename, line, copy);
+                        if (r < 0)
+                                return r;
+                        r = strv_consume(sv, TAKE_PTR(copy));
+                        if (r < 0)
+                                return r;
```

Comment:
should probably log. given that this can only fail on OOM, maybe just use `return log_oom()` here